core/server/worker/src/main/java/alluxio/worker/block/UfsIOManager.java

```diff
@@ -0,0 +1,298 @@
+      .newSingleThreadExecutor(ThreadFactoryUtils.build("UfsIOManager-Scheduler-%d", true));
+
+  /**
+   * @param ufsClient ufs client
+   */
+  public UfsIOManager(UfsManager.UfsClient ufsClient) {
+    mUfsClient = ufsClient;
+    mScheduleExecutor.submit(this::schedule);
+  }
+
+  /**
+   * close.
+   */
+  @Override
+  public void close() {
+    mScheduleExecutor.shutdownNow();
+    mUfsIoExecutor.shutdownNow();
+  }
+
+  /**
+   * set throughput quota for user.
+   * @param user the client name or tag
+   * @param throughput throughput limit, MB/s
+   */
+  public void setQuota(String user, int throughput) {
+    Preconditions.checkArgument(throughput > 0, "throughput should be positive");
+    mThroughputQuota.put(user, throughput);
+  }
+
+  private void schedule() {
+    while (!Thread.currentThread().isInterrupted()) {
+      ReadTask task = null;
+
+      try {
+        task = mReadQueue.take();
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+      if (task != null) {
+        int quota = mThroughputQuota.getOrDefault(task.getUser(), -1);
+        double rate = getUsedThroughput(task.mUser);
+        if (quota == -1 || quota * Constants.MB > rate) {
+          try {
+            mUfsIoExecutor.submit(task);
+          } catch (RejectedExecutionException e) {
+            // should not reach here since we use unbounded queue in thread pool and use a bounded
+            // blocking queue upfront to control number of read tasks
+            mReadQueue.add(task);
+          }
+        } else { // resubmit to queue
+          mReadQueue.add(task);
+        }
+      }
+    }
+  }
+
+  @VisibleForTesting
+  public double getUsedThroughput(String user) {
+    Meter ufsBytesReadThroughput = mUfsBytesReadThroughputMetrics.computeIfAbsent(
+        new BytesReadMetricKey(mUfsClient.getUfsMountPointUri(), user),
+        key -> key.getUser() == null
+            ? MetricsSystem.meterWithTags(MetricKey.WORKER_BYTES_READ_UFS.getName(),
+            MetricKey.WORKER_BYTES_READ_UFS.isClusterAggregated(), MetricInfo.TAG_UFS,
+            MetricsSystem.escape(key.getUri()))
+            : MetricsSystem.meterWithTags(MetricKey.WORKER_BYTES_READ_UFS.getName(),
+            MetricKey.WORKER_BYTES_READ_UFS.isClusterAggregated(), MetricInfo.TAG_UFS,
+            MetricsSystem.escape(key.getUri()), MetricInfo.TAG_USER, key.getUser()));
+    return ufsBytesReadThroughput.getOneMinuteRate();
+  }
+
+  /**
+   * read from ufs.
+   * 
+   * @param blockId block id
+   * @param blockSize block size
+   * @param ufsPath ufs path
+   * @param offset offset to read
+   * @param length length to read
+   * @param positionShort is position short or not
+   * @param user user tag
+   * @return content read
+   * @throws ResourceExhaustedException any exception from ufs
+   */
+  public CompletableFuture<byte[]> read(long blockId, long blockSize, String ufsPath, long offset,
```

Comment:
Actually here "user" is more like a job ID, and we are tracking bandwidth per job ID. It might be better to just call it `String id` or `String tag`. And in this case your metric key doesn't need to include the alluxio URI, so you won't need toe `BytesReadMetricKey` class.